### PR TITLE
man: udevadm: clarify the behavior when multiple matching rules are specified

### LINKED ISSUE #2995: libudev: documentation should clarify when multiple matches are combined via OR and when via AND
### Submission type
- [x] Bug report
- [ ] Request for enhancement (RFE)
### systemd version the issue has been seen with

> 229-3
### Used distribution

> archlinux with linux kernel 4.4.1-2
### Expected behaviour you didn't see

I am using `udevadm trigger --dry-run --verbose` to enumerate devices from the shell exactly as I would with corresponding `udev_enumerate_*()` libudev functions. 

The returned list is wrong (or maybe the documentation is?) when several matchers are combined (libudev: `udev_enumerate_add_match_*()`): I am expecting that every "matchers" are combined, so that the list of devices matching **every** criteria should be returned. But it does not in the following example.

Given a usb-serial device, I want to get its node (/sys/.../ttyUSB0) by enumerating the list of devices uniquely matching it (subsystem usb AND vendor/product ids AND subsystem tty):

```
$ udevadm trigger --dry-run --verbose --subsystem-match=usb --attr-match=idVendor=0e0f --attr-match=idProduct=0002 -subsystem-match=tty
```

Resulting in `/sys/.../ttyUSB0` in my case:

```
/sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0/tty/ttyUSB0
```
### Unexpected behaviour you saw

But only the usb node is returned:

```
/sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2
```

Either the documentation about how matchers are added is wrong/incomplete (but in my opinion, the "add" suggests the union), or there is a problem in the combination of matchers.

My workaround for now is to rerun it with the `--parent-match` option:

```
$ udevadm trigger --dry-run --verbose --parent-match=$(udevadm trigger --dry-run --verbose --subsystem-match=usb --attr-match=idVendor=0e0f --attr-match=idProduct=0002 --subsystem-match=tty) --subsystem-match=tty
/sys/devices/pci0000:00/0000:00:11.0/0000:02:00.0/usb2/2-2/2-2.1/2-2.1:1.0/ttyUSB0/tty/ttyUSB0
```

### PR DESCRIPTION
**Pull request by yuwata**
Closes #2995.

### CODE REVIEW

File: man/udevadm.xml [hunk 1]
```diff
@@ -226,86 +226,81 @@
...
-            attribute name, the content of the attribute is matched
-            against the given value using shell style pattern
-            matching. If no value is specified, the existence of the
-            sysfs attribute is checked. This option can be specified
-            multiple times.</para>
+            <para>Do not trigger events for devices with a matching sysfs attribute. If a value is specified
+            along with the attribute name, the content of the attribute is matched against the given value
+            using shell style pattern matching. If no value is specified, the existence of the sysfs attribute
+            is checked. When this option is specified multiple times, then each matching result is ANDed,
+            that is, only devices which have no specified attribute are triggered.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: "have none of the specified attributes"

File: man/udevadm.xml [hunk 2]
```diff
@@ -226,86 +226,81 @@
...
         <varlistentry>
           <term><option>-p</option></term>
           <term><option>--property-match=<replaceable>PROPERTY</replaceable>=<replaceable>VALUE</replaceable></option></term>
           <listitem>
-            <para>Trigger events for devices with a matching property
-            value. This option can be specified multiple times and
-            supports shell style pattern matching.</para>
+            <para>Trigger events for devices with a matching property value. This option supports shell style
+            pattern matching. When this option is specified more than once, then each matching result is ORed,
+            that is, devices which have at least one matching properties are triggered.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: "one matching property"

File: man/udevadm.xml [hunk 3]
```diff
@@ -226,86 +226,81 @@
...
           <listitem>
-            <para>Trigger events for devices for which the last component
-            (i.e. the filename) of the <filename>/sys</filename> path matches
-            the specified <replaceable>PATH</replaceable>. This option can be
-            specified multiple times and also supports shell style pattern
-            matching.</para>
+            <para>Trigger events for devices for which the last component (i.e. the filename) of the
+            <filename>/sys</filename> path matches the specified <replaceable>PATH</replaceable>. This option
+            supports shell style pattern matching. When this option is specified more than once, then each
+            matching result is ORed.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: Please add "`, that is, all devices which have any of the specified <replaceable>NAME</> are triggered`" as in other cases.

File: man/udevadm.xml [hunk 4]
```diff
@@ -226,86 +226,81 @@
...
           </listitem>
         </varlistentry>
         <varlistentry>
           <term><option>--name-match=<replaceable>NAME</replaceable></option></term>
           <listitem>
-            <para>Trigger events for devices with a matching
-            device path. This option can be specified multiple
-            times.</para>
+            <para>Trigger events for devices with a matching device path.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: No explanation about multiple matches here?

File: man/udevadm.xml [hunk 5]
```diff
@@ -226,86 +226,81 @@
...
           </listitem>
         </varlistentry>
         <varlistentry>
           <term><option>-b</option></term>
           <term><option>--parent-match=<replaceable>SYSPATH</replaceable></option></term>
           <listitem>
-            <para>Trigger events for all children of a given
-            device.</para>
+            <para>Trigger events for all children of a given device.</para>
```
Comment (keszybz) [inline man/udevadm.xml]: And here?

File: man/udevadm.xml [hunk 6]
```diff
@@ -483,14 +478,18 @@
...
           </listitem>
         </varlistentry>
         <varlistentry>
           <term><option>-t</option></term>
           <term><option>--tag-match=<replaceable>string</replaceable></option></term>
           <listitem>
-            <para>Filter udev events by tag. Only udev events with a given tag attached will pass.</para>
+            <para>Filter udev events by tag. Only udev events with a given tag attached will pass.
+            When this option is specified more than once, then each matching result is ORed, that is, devices which have at least one
+            specified tags are monitored.</para>
```
Comment (keszybz) [inline man/udevadm.xml:495]: "one of the specified"

File: man/udevadm.xml (hunk 4 above)
Comment (yuwata) [reply man/udevadm.xml]: This does not support that. The last devpath is used. See #11478 and #6098.

Comment (keszybz) [reply man/udevadm.xml]: Then I think it it's better to say so ("If this option is specified multiple times, the last value counts." or smthg).

**Comment by yuwata** [comment]
@keszybz Thanks. PTAL.

File: man/udevadm.xml (hunk 2 above)
Comment (yuwata) [reply man/udevadm.xml]: In the updated version, I use slightly different expression.

File: man/udevadm.xml (hunk 4 above)
Comment (yuwata) [reply man/udevadm.xml]: OK.

**Comment by yuwata** [comment]
@keszybz Done. `--name-match` or friends are also updated.

File: man/udevadm.xml [hunk 7]
```diff
@@ -182,11 +182,11 @@
 
     <refsect2><title>udevadm trigger
       <arg choice="opt"><replaceable>options</replaceable></arg>
-      <arg choice="opt" rep="repeat"><replaceable>devpath</replaceable>|<replaceable>file</replaceable>|<replaceable>unit</replaceable></arg>
+      <arg choice="opt"><replaceable>devpath</replaceable>|<replaceable>file</replaceable>|<replaceable>unit</replaceable></arg>
       </title>
       <para>Request device events from the kernel. Primarily used to replay events at system coldplug time.</para>
 
-      <para>Takes one or more device specifications as arguments. See the description of <command>info</command>
+      <para>Takes a device specification as a positional argument. See the description of <command>info</command>
```
Comment (keszybz) [inline man/udevadm.xml:189]: Oh, that's so awful. I guess we can merge this, but let's fix in the future to allow multiple arguments.

**Pull request merged**